tracing-subscriber/src/registry/extensions.rs

```diff
@@ -0,0 +1,194 @@
+// taken from https://github.com/hyperium/http/blob/master/src/extensions.rs.
+
+use crate::sync::{RwLockReadGuard, RwLockWriteGuard};
+use std::{
+    any::{Any, TypeId},
+    collections::HashMap,
+    fmt,
+    hash::{BuildHasherDefault, Hasher},
+};
+
+#[allow(warnings)]
+type AnyMap = HashMap<TypeId, Box<dyn Any + Send + Sync>, BuildHasherDefault<IdHasher>>;
+
+/// With TypeIds as keys, there's no need to hash them. They are already hashes
+/// themselves, coming from the compiler. The IdHasher holds the u64 of
+/// the TypeId, and then returns it, instead of doing any bit fiddling.
+#[derive(Default, Debug)]
+struct IdHasher(u64);
+
+impl Hasher for IdHasher {
+    fn write(&mut self, _: &[u8]) {
+        unreachable!("TypeId calls write_u64");
+    }
+
+    #[inline]
+    fn write_u64(&mut self, id: u64) {
+        self.0 = id;
+    }
+
+    #[inline]
+    fn finish(&self) -> u64 {
+        self.0
+    }
+}
+
+/// An immutable, read-only reference to a Span's extensions.
+#[derive(Debug)]
+pub struct Extensions<'a> {
+    inner: RwLockReadGuard<'a, ExtensionsInner>,
+}
+
+impl<'a> Extensions<'a> {
+    pub(crate) fn new(inner: RwLockReadGuard<'a, ExtensionsInner>) -> Self {
+        Self { inner }
+    }
+
+    /// Immutably borrows a type previously inserted into this `Extensions`.
+    pub fn get<T: 'static>(&self) -> Option<&T> {
+        self.inner.get::<T>()
+    }
+}
+
+/// An mutable reference to a Span's extensions.
+#[derive(Debug)]
+pub struct ExtensionsMut<'a> {
+    inner: RwLockWriteGuard<'a, ExtensionsInner>,
+}
+
+impl<'a> ExtensionsMut<'a> {
+    pub(crate) fn new(inner: RwLockWriteGuard<'a, ExtensionsInner>) -> Self {
+        Self { inner }
+    }
+
+    /// Insert a type into this `Extensions`.
+    ///
+    /// If a extension of this type already exists, it will
+    /// be returned. Note that extensions are _not_
+    /// `Layer`-specific—they are _span_-specific. This means that
+    /// other layers can access and mutate extensions that
+    /// a different Layer recorded. For example, an application might
+    /// have a layer that records execution timings alongside a layer
+    /// that reports spans and events to a distributed
+    /// tracing system that requires timestamps for spans.
+    /// Ideally, if one layer records a timestamp _x_, the other layer
+    /// should be able to reuse timestamp _x_.
+    ///
+    /// Therefore, extensions should generally be newtypes, rather than common
+    /// types like [`String`](std::string::String), to avoid accidnental
```

Comment:
do intra-RustDoc links work on stable yet? if not, can we switch this to a URL?

Reply:
typo:
```suggestion
    /// types like [`String`](std::string::String), to avoid accidental
```

Reply:
It _did_ build on stable, but that might be an unintended MSRV bump. I changed this to link to https://doc.rust-lang.org/std/string/struct.String.html directly.